### PR TITLE
Compute filter rules in the asset-syncer

### PR DESCRIPTION
**Pull request by andresmgot**
### Description of the change

Follow up of #2461

Adapt the `asset-syncer` to filter assets based on the given input.

### Applicable issues

  - ref #1583

### Additional information

Note that the filter is applied to the latest version of the chart (the one considered generic).

### CODE REVIEW

File: cmd/asset-syncer/utils.go [hunk 1]
```diff
@@ -145,6 +148,68 @@ func (r *HelmRepo) Repo() *models.RepoInternal {
 	return r.RepoInternal
 }
 
+func satisfy(chartInput map[string]interface{}, rule *apprepov1alpha1.FilterRuleSpec) (bool, error) {
+	query, err := gojq.Parse(rule.JQ)
+	if err != nil {
+		return false, fmt.Errorf("Unable to parse JQuery: %v", err)
+
```
Comment (absoludity) [inline cmd/asset-syncer/utils.go:155]: Not sure why `gofmt` didn't remove the extra space - don't care, just looked odd.

File: cmd/asset-syncer/utils.go [hunk 2]
```diff
@@ -145,6 +148,68 @@ func (r *HelmRepo) Repo() *models.RepoInternal {
...
+		query,
+		gojq.WithVariables(varNames),
+	)
+	if err != nil {
+		return false, fmt.Errorf("Unable to compile JQuery: %v", err)
+	}
+
+	res, _ := code.Run(chartInput, varValues...).Next()
+	if err, ok := res.(error); ok {
+		return false, fmt.Errorf("Unable to compile JQuery: %v", err)
```
Comment (absoludity) [inline cmd/asset-syncer/utils.go:173]: Unable to *run* (and not JQuery :P)

File: cmd/asset-syncer/utils.go [hunk 3]
```diff
@@ -145,6 +148,68 @@ func (r *HelmRepo) Repo() *models.RepoInternal {
...
+	for name, val := range rule.Variables {
+		varNames = append(varNames, name)
+		varValues = append(varValues, val)
+	}
+	code, err := gojq.Compile(
+		query,
+		gojq.WithVariables(varNames),
+	)
+	if err != nil {
+		return false, fmt.Errorf("Unable to compile JQuery: %v", err)
```
Comment (absoludity) [inline cmd/asset-syncer/utils.go:168]: jq != JQuery

File: cmd/asset-syncer/utils.go [hunk 4]
```diff
@@ -145,6 +148,68 @@ func (r *HelmRepo) Repo() *models.RepoInternal {
...
+	}
+
+	res, _ := code.Run(chartInput, varValues...).Next()
+	if err, ok := res.(error); ok {
+		return false, fmt.Errorf("Unable to compile JQuery: %v", err)
+	}
+
+	satisfied, ok := res.(bool)
+	if !ok {
+		return false, fmt.Errorf("Unable to convert JQuery result to boolean. Got: %v", res)
```
Comment (absoludity) [inline cmd/asset-syncer/utils.go:178]: jq

File: cmd/asset-syncer/utils.go [hunk 5]
```diff
@@ -145,6 +148,68 @@ func (r *HelmRepo) Repo() *models.RepoInternal {
...
+	}
+	code, err := gojq.Compile(
+		query,
+		gojq.WithVariables(varNames),
+	)
+	if err != nil {
+		return false, fmt.Errorf("Unable to compile JQuery: %v", err)
+	}
+
+	res, _ := code.Run(chartInput, varValues...).Next()
```
Comment (absoludity) [inline cmd/asset-syncer/utils.go:171]: So everything above this line in this function has been compiling the jq query, independent of any chartInput. Yet this function is called during an iteration of charts, so we're doing all of the above to compile/prepare the jq query each time, for each chart. It seems odd that we're not compiling the `code` in a separate function and calling that just once below *before* iterating the charts?

File: cmd/asset-syncer/utils_test.go [hunk 6]
```diff
@@ -1169,3 +1181,133 @@ func Test_extractFilesFromBuffer(t *testing.T) {
...
+				{Name: "bar"},
+			},
+			apprepov1alpha1.FilterRuleSpec{
+				JQ: `.name | test($var1)`, Variables: map[string]string{"$var1": ".*oo.*"},
+			},
+			[]models.Chart{
+				{Name: "foo"},
+			},
+			nil,
+		},
```
Comment (absoludity) [inline cmd/asset-syncer/utils_test.go:1298]: Sweet, nice to see how the gojq lib fits our needs.

**Review by absoludity — APPROVED** [review]
I did see this yesterday, but didn't think it was worth looking at until the update for the filter spec format was done. One small possible refactor to consider if you agree, but fine either way. Thanks!

File: cmd/asset-syncer/utils.go (hunk 3 above)
Comment (andresmgot) [reply cmd/asset-syncer/utils.go]: ookay, thought that `jq query` was redundant but it's true that jquery is something else

File: cmd/asset-syncer/utils.go (hunk 5 above)
Comment (absoludity) [reply cmd/asset-syncer/utils.go]: Great, thanks for the update to compile it once before the iteration!

**Pull request merged**